checker/src/main/java/org/checkerframework/checker/resourceleak/MustCallInferenceLogic.java

```diff
@@ -0,0 +1,284 @@
+                  .append(method.getSimpleName().toString())
+                  .toSet();
+          fieldsWithCMVs.remove(fieldsWithCMV);
+          fieldsWithCMVs.add(new FieldWithCMVs(receiverEl, calledMethodsnew));
+        } else {
+          Set<String> calledMethods = new HashSet<>();
+          calledMethods.add(method.getSimpleName().toString());
+          fieldsWithCMVs.add(new FieldWithCMVs(receiverEl, calledMethods));
+        }
+      }
+    }
+  }
+
+  /**
+   * Gets the FieldsWithCMVs whose field is equal to the elt, if one exists in {@code
+   * fieldsWithCMVs}, otherwise returns {@code null}.
+   *
+   * @param fieldsWithCMVs set of FieldWithCMVs
+   * @param elt the field element
+   * @return the FieldWithCMVs in {@code fieldsWithCMVs} whose field is equal to {@code elt}, or
+   *     {@code null} if there is no such FieldWithCMVs
+   */
+  private @Nullable FieldWithCMVs getFieldsWithCMVs(
+      Element elt, Set<FieldWithCMVs> fieldsWithCMVs) {
+    for (FieldWithCMVs fieldsWithCMV : fieldsWithCMVs) {
+      if (fieldsWithCMV.field.equals(elt)) {
+        return fieldsWithCMV;
+      }
+    }
+    return null;
+  }
+
+  /**
+   * Propagates a set of FieldWithCMVs to successors. If the successor is a regular exit point, then
+   * it searches for owning fields by looking at the called methods set stored in FieldWithCMVs and
+   * the must call obligation of each field.
+   *
+   * @param curBlock the current block
+   * @param fieldsWithCMVs the set of FieldWithCMVs for the current block
+   * @param visited block-FieldWithCMVs pairs already on the worklist
+   * @param worklist current worklist
+   */
+  private void propagateRegPaths(
+      BlockWithFieldsFacts curBlock,
+      Set<FieldWithCMVs> fieldsWithCMVs,
+      Set<BlockWithFieldsFacts> visited,
+      Deque<BlockWithFieldsFacts> worklist) {
+
+    Set<Block> successors = curBlock.block.getSuccessors();
+
+    for (Block b : successors) {
+      if (b.getType() == Block.BlockType.SPECIAL_BLOCK) {
+
+        if (((SpecialBlock) b).getSpecialType() == SpecialBlock.SpecialBlockType.EXCEPTIONAL_EXIT) {
+          // Do nothing if the successor block is an exceptional exit point
+          continue;
+        }
+
+        findOwningFields(fieldsWithCMVs, cfgMethod.getMethod());
+      }
+
+      BlockWithFieldsFacts successor = new BlockWithFieldsFacts(b, fieldsWithCMVs);
+      if (visited.add(successor)) {
+        worklist.add(successor);
+      }
+    }
+  }
+
+  /**
+   * Checks all element in the fieldsWithCMVs and adds an {@code @Owning} annotation on a field if
+   * the called methods set of that field in {@link FieldWithCMVs} includes all the methods in
+   * the @MustCall type of the field. It also updates classToFieldToFinalizers with the new detected
+   * finalizer.
+   *
+   * @param fieldsWithCMVs the set of FieldWithCMVs
+   * @param methodTree the method tree
+   */
+  private void findOwningFields(Set<FieldWithCMVs> fieldsWithCMVs, Tree methodTree) {
```

Comment:
We need a better name for this method.  It actually goes in and mutates the `WholeProgramInference` object, which I wouldn't expect from the name.  Maybe `updateOwningFieldsInWPI`?